### ISSUE TITLE
load_state_dict creates reference cycle, is there a python garbage collect operation before raise CUDA out-of-memory error?

### ISSUE THREAD

**Issue by Stonesjtu** [pytorch/pytorch#20199]
## 🐛 Bug

Python's garbage collection invokes every 700 object creations (by default), so there should be a situation when the object is not referred but still occupies the CUDA memory (when reference cycles happen), so I'm asking for an additional garbage collecting operation if the CUDA memory is out.

**Comment by VitalyFedyunin** [comment]
It is possible to register GC callback if CUDA is out of memory. We would gladly accept PR which is doing it.

**Comment by colesbury** [comment]
> Python's garbage collection invokes every 700 object creations (by default), so there should be a situation when the object is not referred but still occupies the CUDA memory

This is **not** typically the case. Python objects including `torch.Tensor` are referenced counted. They are usually collected when the last reference is dropped. The collection is only delayed if they are involved in a reference cycle.

**Comment by Stonesjtu** [comment]
@colesbury But the reference cycles are almost everywhere in complicated (or simple) codebase

In the example below, the memory consumption keeps increasing until reaching CPython's threshold.

But if you uncomment the `gc.collect()`, it remains constant.

```notebook
In [11]: blstm = torch.nn.LSTM(1024, 1024, bidirectional=True)
In [12]: blstm.cuda()
Out[12]: LSTM(1024, 1024, bidirectional=True)
In [13]: fake_input = torch.Tensor(10, 10, 1024).cuda().uniform_()
In [14]: for _ in range(1000):
...:     import copy
...:     import gc
...:     blstm.load_state_dict(copy.deepcopy(blstm).state_dict())
...:     # print(gc.get_stats())
...:     # gc.collect()
...:     print('using memory: {} GB'.format(torch.cuda.memory_allocated() / (2**30)))
using memory: 0.25859832763671875 GB
using memory: 0.38469743728637695 GB
using memory: 0.44725847244262695 GB
using memory: 0.509819507598877 GB
using memory: 0.572380542755127 GB
using memory: 0.634941577911377 GB
using memory: 0.697502613067627 GB
using memory: 0.760063648223877 GB
using memory: 0.822624683380127 GB
using memory: 0.885185718536377 GB
using memory: 0.947746753692627 GB
using memory: 1.010307788848877 GB
using memory: 1.072868824005127 GB
using memory: 1.135429859161377 GB
using memory: 1.197990894317627 GB
using memory: 1.260551929473877 GB
using memory: 1.323112964630127 GB
using memory: 1.385673999786377 GB
using memory: 1.448235034942627 GB
using memory: 1.510796070098877 GB
using memory: 1.573357105255127 GB
using memory: 1.635918140411377 GB
using memory: 1.698479175567627 GB
using memory: 1.761040210723877 GB
using memory: 1.823601245880127 GB
using memory: 1.886162281036377 GB
using memory: 1.948723316192627 GB
```

**Comment by colesbury** [comment]
@Stonesjtu we shouldn't generate any reference cycles during normal operation. For example, we worked hard to ensure that autograd backwards doesn't generate any reference cycles. It's really a bug if we do: relying on the GC can still result in OOM and way too much GPU memory usage.

What you describe in load_state_dict should be fixed. The fix is simple: a `del load` after this line:
https://github.com/pytorch/pytorch/blob/9499c7b7ee9afb2f62c4fd6ac39be9b493ff09cc/torch/nn/modules/module.py#L763

For historical context, we tried what you suggested way back in Lua Torch days:

https://github.com/torch/torch7/pull/280
https://github.com/torch/cutorch/pull/219

I'd much rather work to ensure that we don't generate reference cycles within the PyTorch library.
